File: drivers/dht22Sensor.go
package drivers

import (
	"fmt"
	"time"

	"airguard/config"
	"airguard/lib"
	"tinygo.org/x/drivers/dht"
)

var (
	sensor1 dht.Device
	sensor2 dht.Device
)

func InitDHT22() {
	sensor1 = dht.New(config.DHT1Pin, dht.DHT22)
	sensor2 = dht.New(config.DHT2Pin, dht.DHT22)
}

func ReadDHT22() (temp1, hum1, temp2, hum2 float32) {
	t1, h1, err1 := sensor1.Measurements()
	if err1 != nil {
		lib.LogFail("DHT22 #1 (GPIO4)", err1.Error())
	} else {
		lib.LogOK("DHT22 #1 (GPIO4)", fmt.Sprintf("temp=%.1f°C hum=%.1f%%", float32(t1)/10, float32(h1)/10))
	}

	time.Sleep(2 * time.Second)

	t2, h2, err2 := sensor2.Measurements()
	if err2 != nil {
		lib.LogFail("DHT22 #2 (GPIO15)", err2.Error())
	} else {
		lib.LogOK("DHT22 #2 (GPIO15)", fmt.Sprintf("temp=%.1f°C hum=%.1f%%", float32(t2)/10, float32(h2)/10))
	}

	return float32(t1) / 10, float32(h1) / 10, float32(t2) / 10, float32(h2) / 10
}

File: src/dht22Sensor.cpp
#include "dht22Sensor.h"
#include "logManagement.h"
#include "pins.h"
#include <DHTesp.h>

static DHTesp sensor1;
static DHTesp sensor2;

void initDHT22() {
    sensor1.setup(DHT1_PIN, DHTesp::DHT22);
    sensor2.setup(DHT2_PIN, DHTesp::DHT22);
}

static void readSensor(DHTesp& sensor, const char* label, float& temp, float& hum) {
    TempAndHumidity data = sensor.getTempAndHumidity();
    if (sensor.getStatus() != DHTesp::ERROR_NONE) {
        logFail(label, sensor.getStatusString());
        temp = 0;
        hum  = 0;
    } else {
        char detail[64];
        snprintf(detail, sizeof(detail), "temp=%.1f°C hum=%.1f%%", data.temperature, data.humidity);
        logOK(label, detail);
        temp = data.temperature;
        hum  = data.humidity;
    }
}

void readAllDHT22(float& temp1, float& hum1, float& temp2, float& hum2) {
    readSensor(sensor1, "DHT22 #1 (GPIO4)",  temp1, hum1);
    delay(2000);
    readSensor(sensor2, "DHT22 #2 (GPIO15)", temp2, hum2);
}
